flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java

```diff
@@ -153,111 +146,215 @@ public T deserialize(T reuse, DataInputView source) throws IOException {
+			// old snapshot case, just compare the type
+			// we don't need to restore any Kryo stuff, since Kryo was never used for persistence,
+			// only for object-to-object copies.
+			final AvroSerializerConfigSnapshot old = (AvroSerializerConfigSnapshot) configSnapshot;
+			return type.equals(old.getTypeClass()) ?
+					CompatibilityResult.compatible() : CompatibilityResult.requiresMigration();
+		}
+		else {
+			return CompatibilityResult.requiresMigration();
 		}
 	}
 
-	// --------------------------------------------------------------------------------------------
+	// ------------------------------------------------------------------------
+	//  Utilities
+	// ------------------------------------------------------------------------
+
+	@Override
+	public TypeSerializer<T> duplicate() {
+		return new AvroSerializer<>(type);
+	}
 
 	@Override
 	public int hashCode() {
-		return 31 * this.type.hashCode() + this.typeToInstantiate.hashCode();
+		return 42 + type.hashCode();
 	}
 
 	@Override
 	public boolean equals(Object obj) {
-		if (obj instanceof AvroSerializer) {
-			@SuppressWarnings("unchecked")
-			AvroSerializer<T> avroSerializer = (AvroSerializer<T>) obj;
-
-			return avroSerializer.canEqual(this) &&
-				type == avroSerializer.type &&
-				typeToInstantiate == avroSerializer.typeToInstantiate;
-		} else {
+		if (obj == this) {
+			return true;
+		}
+		else if (obj != null && obj.getClass() == AvroSerializer.class) {
+			final AvroSerializer that = (AvroSerializer) obj;
+			return this.type == that.type;
+		}
+		else {
 			return false;
 		}
 	}
 
 	@Override
 	public boolean canEqual(Object obj) {
-		return obj instanceof AvroSerializer;
+		return obj.getClass() == this.getClass();
 	}
 
-	// --------------------------------------------------------------------------------------------
-	// Serializer configuration snapshotting & compatibility
-	// --------------------------------------------------------------------------------------------
-
 	@Override
-	public AvroSerializerConfigSnapshot<T> snapshotConfiguration() {
-		return new AvroSerializerConfigSnapshot<>(type, typeToInstantiate, kryoRegistrations);
+	public String toString() {
+		return getClass().getName() + " (" + getType().getName() + ')';
 	}
 
-	@SuppressWarnings("unchecked")
-	@Override
-	public CompatibilityResult<T> ensureCompatibility(TypeSerializerConfigSnapshot configSnapshot) {
-		if (configSnapshot instanceof AvroSerializerConfigSnapshot) {
-			final AvroSerializerConfigSnapshot<T> config = (AvroSerializerConfigSnapshot<T>) configSnapshot;
+	// ------------------------------------------------------------------------
+	//  Initialization
+	// ------------------------------------------------------------------------
+
+	private void checkAvroInitialized() {
+		if (writer == null) {
+			initializeAvro();
+		}
+	}
+
+	private void initializeAvro() {
+		final ClassLoader cl = Thread.currentThread().getContextClassLoader();
```

Comment:
This is the pattern used by the Kryo serializer for a while rather than `type.getClass().getClassLoader()`, because `type` could in theory be a collection class (application class loader) containing the actual type from another class loader.

Might not be possible to happen for Avro, through, not totally sure...